frontend/src/metabase/lib/expressions/typechecker.js

```diff
@@ -47,18 +47,15 @@ export function typeCheck(cst, startRule) {
       return super.metricExpression(ctx);
     }
 
-    segmentExpression(ctx) {
-      const type = this.typeStack[0];
-      if (type !== "boolean") {
-        throw new Error("Incorrect type for segment");
-      }
-      return super.segmentExpression(ctx);
-    }
-
     dimensionExpression(ctx) {
       const type = this.typeStack[0];
-      if (type === "boolean" || type === "aggregation") {
-        throw new Error("Incorrect type for dimension");
+      if (type === "boolean") {
+        ctx.resolveAs = "segment";
+      } else {
+        ctx.resolveAs = "dimension";
+        if (type === "aggregation") {
+          throw new Error("Incorrect type for dimension");
+        }
```

Comment:
I'm not sure I understand this. It's probably due to gaps in my MQBL/QP knowledge.

Why can't a dimension be used within a boolean context? Would a boolean field not be considered a dimension?

Reply:
It definitely should! Unfortunately right now it's not possible because the grammar enforces the type constraints at the syntax level. Hence, the whole idea of postponing the resolving of an identifier. This is however only possible at the later stage (not this PR, since this change only reaffirms the current behavior without any grammar modification).